android/app/src/main/java/app/organicmaps/MwmActivity.java

```diff
@@ -2004,6 +2064,61 @@ public void onSearchRoutePoint(@RoutePointInfo.RouteMarkType int pointType)
   @Override
   public void onRoutingStart()
   {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
+    {
+      if (!mNavigationBatterySaverPermission &&
+          !mPowerManager.isIgnoringBatteryOptimizations(getPackageName()))
+      {
+        Logger.w(TAG, "Checking for App battery optimisation permission");
+        mNavigationBatterySaverPermission = true;
+        dismissAlertDialog();
+        final MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(this, R.style.MwmTheme_AlertDialog)
+            .setTitle(R.string.battery_optimisation_dialog_title)
+            .setCancelable(false)
+            .setMessage(R.string.battery_optimisation_dialog_summary)
+            .setNegativeButton(R.string.deny, (dialog, which) -> onRoutingStart())
+            .setOnDismissListener(dialog -> mAlertDialog = null);
+
+        final Intent intent = Utils.makeAppBatteryOptimizationIntent(this);
+        if (intent != null)
+        {
+          builder.setPositiveButton(getString(R.string.allow), (dlg, which) -> {
+            mBatterySaverPermission.launch(intent);
+          });
+        }
+        builder.show();
+        return;
+      }
+    }
+
+    // Check for device battery saver mode
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
+    {
+      if (!mNavigationDeviceBatterySaverPermission &&
+          mPowerManagerCompat.isPowerSaveMode(this))
+      {
+        Logger.w(TAG, "Checking for Device battery saver permission");
```

Comment:
Yess..after showing our dialog we are requesting permission and starting a separate activity through intent. 

If we will hide our dialog then we will not be able to tell users why disabling power optimisation is necessary. 

Should i remove the dialog still?

Reply:
Do we want BOTH Utils.makePowerSaverSettingIntent(this) and mBatterySaverPermission.launch(intent) or anyone is enough for us?

I am not sure if i am currently understanding or not, but if you mean do we need both Utils.makePoweSaverSettingIntent(this) and mBatterSaverPermission.launch(intent)

Then yes because Utils.makePoweSaverSettingIntent(this) this class returns an intent if a corresponding activity exists or null if nothing exists.

Where as mBatterSaverPermission.launch(intent) this launches that activity and returns results in callback method.

Please correct me if i misunderstood your question.

Reply:
It is not in general necessary that user have enabled both power saving and optimisation on app and also we are asking the permission only once so i think we can ask both because any of it may become problem for us.

Power saving is the main villain because it is the one which causes issues with satellite fixes so it would be better if we ask user to disable it. battery optimisation for app is not that much dangerous because till now we have't noticed any app suspension by system.

What i am thinking to implement is - 
Asking user only once to disable battery saver and only once to disable the battery optimisation after the installation (similar to starva and mapmyride). Either user allows or denies we will never be asking them again.